### ISSUE TITLE
SQLServer decimal(26,0) causes PANIC in /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tiberius-0.5.9/src/tds/numeric.rs:288:47

### ISSUE THREAD

**Issue by mikeburgh** [prisma/prisma#5950]
Hi Prisma Team! My Prisma Client just crashed. This is the report:
## Versions

| Name            | Version            |
|-----------------|--------------------|
| Node            | v14.15.4           | 
| OS              | debian-openssl-1.1.x|
| Prisma Client   | 2.18.0             |
| Query Engine    | query-engine da6fafb57b24e0b61ca20960c64e2d41f9e8cff1|
| Database        | undefined|

Error:
Given exponent overflowing the maximum accepted scale (255).: TryFromIntError(())

## Query
```
mutation {
  createOneTest(data: {
    id: 5e+22
  }) {
    id
  }
}
```

## Logs
```
ide closed
  prisma:engine       at Socket.onSocketEnd (/home/mike/Dev/prisma-nexus/node_modules/@prisma/client/runtime/index.js:25190:25)
  prisma:engine       at Socket.emit (events.js:327:22)
  prisma:engine       at endReadableNT (internal/streams/readable.js:1327:12)
  prisma:engine       at processTicksAndRejections (internal/process/task_queues.js:80:21) {
  prisma:engine     code: 'UND_ERR_SOCKET'
  prisma:engine   }
  prisma:engine }  
  prisma:engine {
  prisma:engine   error: Error: connect ECONNREFUSED 127.0.0.1:38331
  prisma:engine       at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16) {
  prisma:engine     errno: -111,
  prisma:engine     code: 'ECONNREFUSED',
  prisma:engine     syscall: 'connect',
  prisma:engine     address: '127.0.0.1',
  prisma:engine     port: 38331
  prisma:engine   }
  prisma:engine }  
  prisma:engine { cwd: '/home/mike/Dev/prisma-nexus/prisma' }  
  prisma:engine Search for Query Engine in /home/mike/Dev/prisma-nexus/node_modules/.prisma/client  
  plusX Execution permissions of /home/mike/Dev/prisma-nexus/node_modules/.prisma/client/query-engine-debian-openssl-1.1.x are fine  
  prisma:engine {
  prisma:engine   flags: [
  prisma:engine     '--enable-experimental=microsoftSqlServer',
  prisma:engine     '--enable-raw-queries',
  prisma:engine     '--port',
  prisma:engine     '39157'
  prisma:engine   ]
  prisma:engine }  
  prisma:engine stdout  Starting a mssql pool with 25 connections.  
  prisma:engine stdout  TLS encryption is not enabled. All traffic including the login credentials are not encrypted.  
  prisma:engine stdout  Database change from 'test' to 'master'  
  prisma:engine stdout  Changed database context to 'test'.  
  prisma:engine stdout  SQL collation change from None to windows-1252/windows-1252  
  prisma:engine stdout  Microsoft SQL Server version 4110876686  
  prisma:engine stdout  Packet size change from '4096' to '4096'  
  prisma:engine stdout  Started http server on http://127.0.0.1:39157  
  prisma:engine Search for Query Engine in /home/mike/Dev/prisma-nexus/node_modules/.prisma/client  
  plusX Execution permissions of /home/mike/Dev/prisma-nexus/node_modules/.prisma/client/query-engine-debian-openssl-1.1.x are fine  
  prisma:engine Client Version: 2.18.0  
  prisma:engine Engine Version: query-engine da6fafb57b24e0b61ca20960c64e2d41f9e8cff1  
  prisma:engine Active provider: sqlserver  
  prisma:engine stdout  Begin transaction  
  prisma:engine stdout  PANIC in /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tiberius-0.5.9/src/tds/numeric.rs:288:47
Given exponent overflowing the maximum accepted scale (255).: TryFromIntError(())  
  prisma:engine TypeError: this.currentRequestPromise.cancel is not a function
  prisma:engine     at NodeEngine.handlePanic (/home/mike/Dev/prisma-nexus/node_modules/@prisma/client/runtime/index.js:26934:36)
  prisma:engine     at NodeEngine.setError (/home/mike/Dev/prisma-nexus/node_modules/@prisma/client/runtime/index.js:26880:16)
  prisma:engine     at LineStream.<anonymous> (/home/mike/Dev/prisma-nexus/node_modules/@prisma/client/runtime/index.js:27170:24)
  prisma:engine     at LineStream.emit (events.js:315:20)
  prisma:engine     at addChunk (internal/streams/readable.js:309:12)
  prisma:engine     at readableAddChunk (internal/streams/readable.js:284:9)
  prisma:engine     at LineStream.Readable.push (internal/streams/readable.js:223:10)
  prisma:engine     at LineStream.Transform.push (internal/streams/transform.js:166:32)
  prisma:engine     at LineStream._pushBuffer (/home/mike/Dev/prisma-nexus/node_modules/@prisma/client/runtime/index.js:24055:19)
  prisma:engine     at LineStream._transform (/home/mike/Dev/prisma-nexus/node_modules/@prisma/client/runtime/index.js:24049:10)  {"timestamp":"Mar 02 13:02:25.858","level":"ERROR","fields":{"message":"PANIC","reason":"Given exponent overflowing the maximum accepted scale (255).: TryFromIntError(())","file":"/root/.cargo/registry/src/github.com-1ecc6299db9ec823/tiberius-0.5.9/src/tds/numeric.rs","line":288,"column":47},"target":"query_engine","span":{"name":"execute_single_operation"},"spans":[{"name":"execute_single_operation"}]}  
  prisma:engine {
  prisma:engine   error: SocketError: other side closed
  prisma:engine       at Socket.onSocketEnd (/home/mike/Dev/prisma-nexus/node_modules/@prisma/client/runtime/index.js:25190:25)
  prisma:engine       at Socket.emit (events.js:327:22)
  prisma:engine       at endReadableNT (internal/streams/readable.js:1327:12)
  prisma:engine       at processTicksAndRejections (internal/process/task_queues.js:80:21) {
  prisma:engine     code: 'UND_ERR_SOCKET'
  prisma:engine   }
  prisma:engine }  
```

## Client Snippet
```js
const { PrismaClient } = require('@prisma/client');

const prisma = new PrismaClient();

async function main() {
	const test = await prisma.test.create({
		data: {
			id: 17661757261711787211853,
		},
	});
	console.log(post);

	const allTests = await prisma.test.findMany();
	console.log(allTests);
}

main()
	.catch((e) => {
		throw e;
	})
	.finally(async () => {
		await prisma.$disconnect();
	});

```

## Schema
```prisma
generator client {
    provider        = "prisma-client-js"
    previewFeatures = ["microsoftSqlServer"]
}
datasource db {
    provider = "sqlserver"
    url      = env("DATABASE_URL")
}

model Test {
  id	   Decimal   @id @db.Decimal(26, 0)
}
```

A working repo to reproduce this can be found here; 
https://github.com/mikeburgh/prisma-sqlserver-decimal

**Comment by mikeburgh** [comment]
If it helps, I think the issue lies in the generated code.. it is triggering a rollback and error in SQLServer:

```
prisma:query BEGIN TRAN
prisma:query DECLARE @generated_keys table([id] DECIMAL(32,16)) INSERT INTO [dbo].[Test] ([id]) OUTPUT [Inserted].[id] INTO @generated_keys VALUES (@P1) SELECT [t].[id] FROM @generated_keys AS g INNER JOIN [dbo].[Test] AS [t] ON [t].[id] = [g].[id] WHERE @@ROWCOUNT > 0
prisma:info  Rollback transaction
prisma:query ROLLBACK
```

Cleaned up to run in SQL
```sql
sp_executesql N'
DECLARE @generated_keys table([id] DECIMAL(32,16)) 
INSERT INTO [dbo].[Test] ([id]) 
OUTPUT [Inserted].[id] INTO @generated_keys VALUES (@P1)
SELECT [t].[id] FROM @generated_keys AS g 
INNER JOIN [dbo].[Test] AS [t] ON [t].[id] = [g].[id] 
WHERE @@ROWCOUNT > 0',N'@P1 numeric(26,0)',@P1=21474834921812847092396
```

If you change that decimal definition in the @generated_keys table definition to match the schema, it works:

```sql
sp_executesql N'
DECLARE @generated_keys table([id] DECIMAL(26,0)) 
INSERT INTO [dbo].[Test] ([id]) 
OUTPUT [Inserted].[id] INTO @generated_keys VALUES (@P1)
SELECT [t].[id] FROM @generated_keys AS g 
INNER JOIN [dbo].[Test] AS [t] ON [t].[id] = [g].[id] 
WHERE @@ROWCOUNT > 0',N'@P1 numeric(26,0)',@P1=21474834921812847092396
```

**Comment by pantharshit00** [comment]
I can confirm this, we need to adjust the conversion code in the driver. Thanks for the detailed reproduction,

**Comment by pantharshit00** [comment]
cc @pimeys as it is related to tiberius

**Comment by pimeys** [comment]
It's actually related to Quaint. The error is in the SQL:

```sql
DECLARE @generated_keys table([id] DECIMAL(32,16)) 
```
then later in the query:
```sql
,N'@P1 numeric(26,0)',@P1=21474834921812847092396
```

The numeric is too big to fit to the list `@generated_keys`.

**Comment by pimeys** [comment]
So, SQL Server has a few things in the query building that needs to know a bit about the schema to operate correctly. One is the emulation of `INSERT ... ON CONFLICT DO NOTHING` that turns into a `MERGE` and needs to understand about the underlying unique constraints.

The other one is `RETURNING`, that's in SQL Server `OUTPUT`. We could just use `OUTPUT`, but if the database uses any triggers, a plain `OUTPUT` would fail. The commonly accepted workaround is to declare a list variable, directing the output into it and return the data from there.

What we've done originally here is we just store the type family to the query, and hope for the best. Now to solve the issue here we'd need to actually store more information of the underlying type to the query and use `DECIMAL(20,0)` in the list variable.

@mikeburgh can you find a workaround for now to your issue? I'm having extremely packed sprint and I'm afraid I don't have time to look at this until next sprint starts in bit over a week. I know already the solution for this bug is a bit more involved...

**Comment by mikeburgh** [comment]
@pimeys I still working on getting Prisma integrated into this project.. so all good with next sprint. Appreciate you looking into it!

**Comment by pimeys** [comment]
I mean, this is most definitely fixable. The problem is how the client kind of doesn't know the native types, just the generic prisma types right now. We could hack the decimal data into it, but I'm not sure will there be other types that cause similar problems like this.

Here's the current types for the `INSERT`: https://github.com/prisma/quaint/blob/master/src/visitor/mssql.rs#L49-L61

It maps the situation before native types, but I'd see a problem on e.g. data using `NVARCHAR(MAX)`...

**Comment by mikeburgh** [comment]
Makes sense.. and yes I think you could find other types that cause issues if that list from quaint is the full list currently used for the table variable field mapping.. MAX for one or even longer 2000+ NVARCHAR fields... although I know absolutely zero rust, so I may be miss understanding exactly how that function is operating.

I also have another issue, that's potentially related to this, you can see it in.. #5952. It's not a size of the decimal problem, but rather something with how Prisma re assigns results after it groups like queries into one when it receives a batch to solve the N+1 problem of graphQL. It's also not provider specific.

Again, neither of these are critical or urgent, we are just in the very early stages of adding Prisma into an existing project.

**Comment by pimeys** [comment]
Yeah, the function I linked also follows the standard Prisma types, meaning when we build a query, we just give a notion of the Prisma type of that column, which is definitely not enough to this query.

The second problem you linked is not really related to this.

This is kind of critical, because before we can say SQL Server connector is GA, we had two problems to solve:

- Detection of cyclical references in the data model and a syntax to define cascading rules (important for many SQL Server use-cases).
- An ability to define partial index rules in the data model. A one special case: should we have null values in the index or not.

Now we have three...

**Comment by pimeys** [comment]
Sorry this took a bit longer than usual. We had a few really busy sprints and the ticket got dropped from them.

It's a bit trickier issue that I solved in the pull request above. Some caveats apply... Due to the graphql layer underneath, passing a number here is handled as a float always by the graphql library. This is super unstable and I recommend either passing the id value by string, or using [decimal.js](https://www.npmjs.com/package/decimal.js):

```javascript
const test = await prisma.test.create({
    data: {
        id: "17661757261711787211853",
    },
});
```

If you pass that big of a number as a number down to the engine, it'll be converted to a float first, that will not really fit there and the result is a horribly wrong rounding error. Even worse, it will not fit to the decimal anymore when we convert the float value, causing another error.

We're havin a meeting tomorrow where I'm suggesting to not allow floats in field input which is backed by a decimal type in the database to prevent weird errors from happening.

I'm curious btw, if you're still here, why are you using `decimal(26,0)` as a primary key? I've never seen one in action. You can fit a bigger number here compared to `bigint` for sure, but is it the easy ordering compared to `uniqueidentifier` then?

**Comment by mikeburgh** [comment]
@pimeys still here, does it seemed like a good idea at the time count as an answer ? :) It was a decision made about 10+ years ago now.  We needed an index friendly id (GUID's etc were not a good fit), which was not sequential or stepped, and that was big enough that we could generate lots of them very quickly and ensure uniqueness.  Thanks for getting this sorted!

**Comment by pimeys** [comment]
I understand. Been here in the trenches long enough to understand the reasoning... :D

We're aiming to get the sql server support GA hopefully soon, so now catching the last bugs so we can say it's stable.
